### PR TITLE
Add to_ieee754_32 Presto function

### PR DESCRIPTION
**Pull request by ashokku2022**
Resolves https://github.com/facebookincubator/velox/issues/4425

### CODE REVIEW

**Comment by ashokku2022** [comment]
@aditi-pandit please review the PR when you get time. I have made the adjustment after the file split (refactoring PR).

**Comment by ashokku2022** [comment]
fuzzer output from local run

```
I20230418 15:30:42.551452 1518462 ExpressionFuzzer.cpp:974] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230418 15:30:42.551465 1518462 ExpressionFuzzer.cpp:978] to_ieee754_32 6803 100.00% 223252
I20230418 15:30:42.551483 1518462 ExpressionFuzzer.cpp:984] ==============================> Bottom 1 by number of rows processed
I20230418 15:30:42.551497 1518462 ExpressionFuzzer.cpp:986] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230418 15:30:42.551510 1518462 ExpressionFuzzer.cpp:991] to_ieee754_32 6803 100.00% 223252
I20230418 15:30:42.551525 1518462 ExpressionFuzzer.cpp:1004] ==============================> All stats sorted by number of times the function was chosen
I20230418 15:30:42.551538 1518462 ExpressionFuzzer.cpp:1006] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230418 15:30:42.551553 1518462 ExpressionFuzzer.cpp:1010] to_ieee754_32 6803 100.00% 223252
```

**Review by majetideepak — COMMENTED** [review]
@ashokku2022 
Left some comments. On the Presto side, I also see Hex representation.
```
presto> select to_ieee754_32(10.3);
    _col0    
-------------
 41 24 cc cd 
(1 row)
```

**Comment by ashokku2022** [comment]
> @ashokku2022 Left some comments. On the Presto side, I also see Hex representation.
> 
> ```
> presto> select to_ieee754_32(10.3);
>     _col0    
> -------------
>  41 24 cc cd 
> (1 row)
> ```

@majetideepak that is just for display purposes (the hex conversion, maybe to save space) and I have noticed that too but based on documentation the output should be `varbinary` and not hex.

File: velox/functions/prestosql/BinaryFunctions.h [hunk 1]
```diff
@@ -403,4 +404,21 @@ struct ToBigEndian64 {
...
 };
 
+template <typename T>
+struct ToIEEE754_32Function {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varbinary>& result,
+      const arg_type<float>& input) {
+    static constexpr auto resLength = 8 * sizeof(uint32_t);
```
[the remark this answers is not in the document]
Comment (ashokku2022) [reply velox/functions/prestosql/BinaryFunctions.h]: done

File: velox/docs/functions/presto/binary.rst [hunk 2]
```diff
@@ -86,6 +86,12 @@ Binary Functions
 
     Encodes ``binary`` into a hex string representation.
 
+.. function:: to_ieee754_32(real) -> varbinary
+
+    Computes the 32 bits of IEEE 754 single-precision floating-point format of ``real``.
```
[the remark this answers is not in the document]
Comment (ashokku2022) [reply velox/docs/functions/presto/binary.rst]: done

File: velox/functions/prestosql/tests/BinaryFunctionsTest.cpp [hunk 3]
```diff
@@ -583,4 +583,35 @@ TEST_F(BinaryFunctionsTest, toBigEndian64) {
       toBigEndian64((int64_t)-9223372036854775807 - 1));
 }
 
+TEST_F(BinaryFunctionsTest, toIEEE754_32) {
```
[the remark this answers is not in the document]
Comment (ashokku2022) [reply velox/functions/prestosql/tests/BinaryFunctionsTest.cpp]: done

File: velox/functions/prestosql/BinaryFunctions.h [hunk 4]
```diff
@@ -403,4 +404,21 @@ struct ToBigEndian64 {
...
+    result.resize(resLength);
+    unsigned int* ui = const_cast<unsigned int*>(
+        reinterpret_cast<const unsigned int*>(&input));
+    std::bitset<32> bits = std::bitset<32>(*ui);
+    std::string buffer = bits.to_string();
```
[the remark this answers is not in the document]
Comment (ashokku2022) [reply velox/functions/prestosql/BinaryFunctions.h:418]: done

File: velox/functions/prestosql/BinaryFunctions.h [hunk 5]
```diff
@@ -403,4 +404,20 @@ struct ToBigEndian64 {
...
 };
 
+template <typename T>
+struct ToIEEE754Bits32 {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varbinary>& result,
+      const arg_type<float>& input) {
+    static constexpr auto kResLength = 8 * sizeof(uint32_t);
```
Comment (aditi-pandit) [inline velox/functions/prestosql/BinaryFunctions.h:414]: Nit : Rename kResultLength

File: velox/functions/prestosql/tests/BinaryFunctionsTest.cpp [hunk 6]
```diff
@@ -583,4 +583,36 @@ TEST_F(BinaryFunctionsTest, toBigEndian64) {
       toBigEndian64((int64_t)-9223372036854775807 - 1));
 }
 
+TEST_F(BinaryFunctionsTest, toIEEE754Bits32) {
+  const auto toIEEE754Bits32 = [&](std::optional<float> value) {
+    return evaluateOnce<std::string, float>("to_ieee754_32(c0)", value);
+  };
+
+  EXPECT_EQ(std::nullopt, toIEEE754Bits32(std::nullopt));
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/BinaryFunctionsTest.cpp:591]: Please also add tests for 
std::numeric_limits<T>::infinity(), std::numeric_limits<T>::max(), std::numeric_limits<T>::min(), std::numeric_limits<T>::quiet_NaN().

File: velox/functions/prestosql/BinaryFunctions.h (hunk 5 above)
Comment (ashokku2022) [reply velox/functions/prestosql/BinaryFunctions.h]: done

File: velox/functions/prestosql/tests/BinaryFunctionsTest.cpp (hunk 6 above)
Comment (ashokku2022) [reply velox/functions/prestosql/tests/BinaryFunctionsTest.cpp:592]: added

**Review by majetideepak — APPROVED** [review]

**Comment by ashokku2022** [comment]
@mbasmanova could you please review, thanks!

**Comment by aditi-pandit** [comment]
> @mbasmanova could you please review, thanks!

Adding @Yuhta Jimmy also.

**Review by Yuhta — COMMENTED** [review]
Same as #4662, we should consider reusing `bits::toString` and make sure `buffer` is not reallocated for each row

**Comment by aditi-pandit** [comment]
@ashokku2022 : Please rebase your code.

**Comment by ashokku2022** [comment]
@aditi-pandit @Yuhta have rebase and updated the code (just like [to_ieee754_64](https://github.com/facebookincubator/velox/pull/4662)). Please review, thanks!

**Comment by ashokku2022** [comment]
fuzzer test output:

`velox_expression_fuzzer_test  --seed 123456 --duration_sec 60 --logtostderr=1 --minloglevel=0  --only to_ieee754_32  --v=1 2>&1 | tee fuzzertest2.out`

```
I20230504 15:46:39.856527 5610506 ExpressionFuzzer.cpp:1070] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230504 15:46:39.856540 5610506 ExpressionFuzzer.cpp:1074] to_ieee754_32 6547 100.00% 212966
I20230504 15:46:39.856581 5610506 ExpressionFuzzer.cpp:1080] ==============================> Bottom 1 by number of rows processed
I20230504 15:46:39.856596 5610506 ExpressionFuzzer.cpp:1082] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230504 15:46:39.856611 5610506 ExpressionFuzzer.cpp:1087] to_ieee754_32 6547 100.00% 212966
I20230504 15:46:39.856645 5610506 ExpressionFuzzer.cpp:1100] ==============================> All stats sorted by number of times the function was chosen
I20230504 15:46:39.856683 5610506 ExpressionFuzzer.cpp:1102] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230504 15:46:39.856698 5610506 ExpressionFuzzer.cpp:1106] to_ieee754_32 6547 100.00% 212966
```

File: velox/common/base/BitUtil.cpp [hunk 7]
```diff
@@ -90,6 +90,18 @@ void copyBitsBackward(
   }
 }
 
+void toString(const uint32_t* bits, int offset, int size, char* out) {
```
Comment (Yuhta) [inline velox/common/base/BitUtil.cpp:93]: We don't need these 2, just cast to `const uint64_t*`

Comment (ashokku2022) [reply velox/common/base/BitUtil.cpp]: done, and pushed new changes

**Comment by Yuhta** [comment]
Hi @ashokku2022, @mbasmanova pointed out that the semantics of your `to_ieee754_32` and `to_ieee754_64` is incorrect, they should return binary form of the data (32 bits and 64 bits) instead of the string binary (32 bytes and 64 bytes).  Can you correct them?

**Comment by Yuhta** [comment]
The correct value returns in Presto Java:

```
presto:di> select to_ieee754_32(real '10.3');
    _col0
-------------
 41 24 cc cd
(1 row)
```

**Comment by Yuhta** [comment]
@ashokku2022 In the initial version you convert the `bitset` to string and write it to output.  It is the same results as current version.  What Presto Java has is the binary values in the output directly, you just need to use `folly::Endian::big(input)` and write it to output.  All the unit tests need to fixed though.

**Comment by ashokku2022** [comment]
@Yuhta Jimmy, thanks for the feedback, highly appreciate it. Let me work on it and update code. Once this `to_ieee754_32` PR is done, will open another one to address `to_ieee754_64`. Thank you!

**Comment by yzhang1991** [comment]
Hi @ashokku2022, shall we close this since Krishna will add this function in https://github.com/facebookincubator/velox/pull/5336?

**Pull request closed without merging**